drivers/sensor/bq274xx/bq274xx.c

```diff
@@ -90,236 +101,293 @@ static int bq274xx_ctrl_reg_write(const struct device *dev, uint16_t subcommand)
+static int bq274xx_read_block(const struct device *dev,
+			      uint8_t subclass,
+			      uint8_t *block, uint8_t num_bytes)
 {
+	const struct bq274xx_config *const config = dev->config;
 	int ret;
 
-	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_CTRL_DEVICE_TYPE);
+	ret = i2c_reg_write_byte_dt(&config->i2c, BQ274XX_EXT_DATA_CLASS, subclass);
 	if (ret < 0) {
-		LOG_ERR("Unable to write control register");
+		LOG_ERR("Failed to update state subclass");
 		return -EIO;
 	}
 
-	ret = bq274xx_cmd_reg_read(dev, BQ274XX_CMD_CONTROL_LOW, val);
+	/* DataBlock(), 0 for the first 32 bytes. */
+	ret = i2c_reg_write_byte_dt(&config->i2c, BQ274XX_EXT_DATA_BLOCK, 0x00);
 	if (ret < 0) {
-		LOG_ERR("Unable to read register");
+		LOG_ERR("Failed to update block offset");
+		return -EIO;
+	}
+
+	k_sleep(BQ274XX_SUBCLASS_DELAY);
+
+	ret = i2c_burst_read_dt(&config->i2c, BQ274XX_EXT_BLKDAT_START, block, num_bytes);
+	if (ret < 0) {
+		LOG_ERR("Unable to read block data");
 		return -EIO;
 	}
 
 	return 0;
 }
 
-static int bq274xx_gauge_configure(const struct device *dev)
+static int bq274xx_write_block(const struct device *dev,
+			       uint8_t *block, uint8_t num_bytes)
 {
 	const struct bq274xx_config *const config = dev->config;
-	struct bq274xx_data *data = dev->data;
-	const struct bq274xx_regs *regs = data->regs;
+	uint8_t checksum = 0;
 	int ret;
-	uint8_t tmp_checksum, checksum_old, checksum_new;
-	uint16_t flags, designenergy_mwh, taperrate, reg_val;
-	uint8_t block[BQ27XXX_DM_SZ];
-	uint8_t try;
 
-	designenergy_mwh = (uint32_t)config->design_capacity * 37 / 10; /* x3.7 */
-	taperrate = config->design_capacity * 10 / config->taper_current;
-
-	/* Unseal the battery control register */
-	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_UNSEAL_KEY_A);
+	ret = i2c_burst_write_dt(&config->i2c, BQ274XX_EXT_BLKDAT_START, block, num_bytes);
 	if (ret < 0) {
-		LOG_ERR("Unable to unseal the battery");
+		LOG_ERR("Unable to write block data");
 		return -EIO;
 	}
 
-	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_UNSEAL_KEY_B);
+	for (uint8_t i = 0; i < num_bytes; i++) {
+		checksum += block[i];
+	}
+	checksum = 0xff - checksum;
+
+	ret = i2c_reg_write_byte_dt(&config->i2c, BQ274XX_EXT_CHECKSUM, checksum);
 	if (ret < 0) {
-		LOG_ERR("Unable to unseal the battery");
+		LOG_ERR("Failed to update block checksum");
 		return -EIO;
 	}
 
-	/* Send CFG_UPDATE */
-	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_CTRL_SET_CFGUPDATE);
+	k_sleep(BQ274XX_SUBCLASS_DELAY);
+
+	return 0;
+}
+
+static void bq274xx_update_block(uint8_t *block,
+				 uint8_t offset, uint16_t val,
+				 bool *block_modified)
+{
+	uint16_t old_val;
+
+	old_val = sys_get_be16(&block[offset]);
+
+	LOG_DBG("update block: off=%d old=%d new=%d\n", offset, old_val, val);
+
+	if (val == old_val) {
+		return;
+	}
+
+	sys_put_be16(val, &block[offset]);
+
+	*block_modified = true;
+}
+
+static int bq274xx_change_mode(const struct device *dev, uint8_t mode)
```

Comment:
While debugging I noticed this function has some issues.

Here we're actually passing in a 16-bit command:
```suggestion
static int bq274xx_change_mode(const struct device *dev, uint16_t cmd)
```

The function should wait for bit 4 (`BQ27XXX_FLAG_CFGUP`) to set after sending command `BQ274XX_CTRL_SET_CFGUPDATE`

and should wait for bit 4 to clear after sending command `BQ274XX_CTRL_SOFT_RESET`.

Reply:
Oh very interesting, the Linux driver has this bug as well. Ok I've reworked this whole thing to check the bit correctly, I think it's doing it now, I can see it waits for 46 cycles on setup, good stuff.